geoportal/c2cgeoportal_geoportal/lib/functionality.py

```diff
@@ -55,12 +56,22 @@ def _get_config_functionality(name, registered, types, request, errors):
     return [r for r in result if r is not None]
 
 
-def _get_db_functionality(name, role, types, request, errors):
-    result = [
-        get_typed(name, functionality.value, types, request, errors)
-        for functionality in role.functionalities
-        if functionality.name == name
-    ]
+def _get_db_functionality(name, user: static.User, types, request, errors):
+    if types[name].get('single', False):
+        result = [
+            get_typed(name, functionality.value, types, request, errors)
+            for functionality in user.settings_role.functionalities
+            if functionality.name == name
+        ]
+        return [r for r in result if r is not None]
+
+    result = []
+    for role in user.roles:
+        result.extend([
+            get_typed(name, functionality.value, types, request, errors)
+            for functionality in role.functionalities
+            if functionality.name == name
+        ])
```

Comment:
You may have many times the same functionnality with same value:
```
result.extend([
        get_typed(name, functionality.value, types, request, errors)
        for functionality in set(
            [f for role in roles for f in role.functionalities if f.name == name]
        )
    ])
```

Reply:
or maybe better : 
```
result.extend([
        get_typed(name, value, types, request, errors)
        for value in set(
            [f.value for role in roles for f in role.functionalities if f.name == name]
        )
    ])
```

Reply:
You try to find the more complex solution? For me we should avoid to use `[ ... for ... for ...]`, it's too complex to understand ...

Reply:
Ok, so we might create a variable for the values
```
values = set()
for role in roles:
    values = values | set([f.value for f in role.functionalities])
result.extend([
    get_typed(name, value, types, request, errors)
    for value in values
])

Reply:
This might be more readable:
```
    result.extend([
        get_typed(name, value, types, request, errors)
        for value in set(
            [f.value
             for role in roles
             for f in role.functionalities
             if f.name == name]
        )
    ])
```

Reply:
Updated with another more readable possibility...